app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -74,6 +91,60 @@ def save_inventory_multi(association, hashes, deletes, find_key, child_keys = []
     association.push(new_records)
   end
 
+  # Saves a inventory record, and any sub inventory records, without
+  # associating it with the parent object that is passed in.  Parent usually is
+  # an EMS, but doesn't have to be.
+  #
+  # @param parent [ActiveRecord] base record being saved
+  # @param association [Symbol]  associated record type being added to parent
+  # @param hashes [Array]        inventory data (hash per record)
+  # @param find_key [Array]      ordered keys for existing record lookup
+  # @param child_keys [Array]    keys from hashes that are child records
+  # @param extra_keys [Array]    hash keys to be removed before saving records
+  # @param disconnect [Boolean]  whether or not to delete or disconnect records
+  #                              that don't exist in the inventory hashes, but
+  #                              are present in the DB
+  #
+  # @return [Hash] the hashes passed into the method
+  def save_inventory_with_thin_association(parent, association, hashes, find_key,
+                                           child_keys = [], extra_keys = [],
+                                           disconnect = false)
+
+    relation      = parent.association(association).reflection
+    model         = parent.association(association).klass
+    assoc_attrs   = parent.association(association).reader.scope_attributes
+    deletes_index = model.pluck(relation.active_record_primary_key)
+                         .index_by { |x| x }
+
+    child_keys = Array.wrap(child_keys)
+    remove_keys = Array.wrap(extra_keys) + child_keys
+    record_index = IdTypedIndex.new(deletes_index.values, find_key, model)
+    inventory_ids = []
+
+    hashes.each do |h|
+      found = nil
+      ActiveRecord::Base.transaction do
+        found = save_inventory_with_findkey(model, h.except(*remove_keys).merge(assoc_attrs), deletes_index, nil, record_index, true)
+        save_child_inventory(found, h, child_keys)
+        found.save
```

Comment:
if you do found.save here, I think you can just omit the ```store_ids_for_record_ids```

just do
```h[:id] = found.id```

right?

Reply:
Also we already call update_attributes inside. This will not cause another query, since rails check that, but it still calls something to figure out if the record changed again. But maybe this doesn't cost much comparing to other operations.

Reply:
> Also we already call update_attributes inside. This will not cause another query...

Yeah, this was just me copying what existed in `save_inventory_multi`, and I didn't want to remove it incase there was something that was not tested and this `.save` had some reason for being there.  You are probably right, but I was just playing it safe.

> just do
> `h[:id] = found.id`
> 
> right?

Good call.  I think paired with what you said in https://github.com/ManageIQ/manageiq/pull/15989/files#r139679218, I can probably get rid of that `:id => [1,2,3]` bit and change it to `:id => parent.send(:association).select(:id)` instead and avoid a nasty query, or maybe avoid a `.where` all together if a relation is passed in instead of an array.  Will have to do some testing.

Reply:
Right so the reason of save being after save_child_inventory, is that it saves all the ```association << association.build(xy)``` of the child relation. If we would switch to doing create! or update_attributes! inside, we won't need it.

If you wound't have the save here, it would be saved as part o the parent.save above (since all of them were parent associations)

Reply:
@Ladas sometimes you save the parent before the children, but other times you need to save the children before the parent. And in those cases, we have to circle back and set the ids.

I'm sorry I can't recall the exact cases. It is possible that these have been fixed and removed.

Also, a few places I wanted to remove the `save` but it broke tests (I think around updates).
The code has hopefully changed enough to finally allow us to remove the non-optimal saves.

Reply:
Mmk, with that said, I might just keep the `found.save` in for now, and circle back to see if we can remove it in another PR.  Would rather have a "we are confident this shouldn't break anything" PR, and a follow up one to add some further optimizations then try and cram it into a single PR.

Reply:
Sounds like we can do it in baby steps...keep it as is, and then figure out if we can optimize it after.